### ISSUE TITLE
Write some unit tests for Deferreds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4300]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#4300|
|Type|task|
|Created|2010-02-23 15:20:09Z|
|Branch|https://github.com/twisted/twisted/tree/deferred-unit-tests-4300-2|

Deferreds need more unit tests, the current test suite has far too little coverage, making it hard for us to make improvements like [#4593](https://github.com/twisted/twisted/issues/4593) or C implementations of Deferreds.

<details><summary>Searchable metadata</summary>

```
trac-id__4300 4300
type__task task
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_deferred_unit_tests_4300_2 branches/deferred-unit-tests-4300-2
branch_author__exarkun__jml exarkun, jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1266938409000000 1266938409000000
changetime__1279053013000000 1279053013000000
version__None None
owner__ 
cc__jesstess cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [28457]) Branching to 'deferred-unit-tests-4300'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

The branch adds a bunch of unit tests for Deferred, particularly for some thorny cases around Deferreds that have callbacks that return Deferreds that themselves are buried somewhere deep in the callback chain.

The branch also records chaining explicitly, thus allowing us to test and improve the Deferred `repr()`, which makes debugging slightly easier.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess set owner to @jml|
|-|-|

Thanks for the well-documented test cases - they are so clear we could be pointing people looking for more examples of using Deferreds to these. A few comments:

 * copyright bump on test_defer.py
 * I'm having trouble parsing the ```test_asynchronousImplicitErrorChain``` docstring. I think an article and some commas are missing?
 * in ```test_nestedAsynchronousChainedDeferredsWithExtraCallbacks```: "Create a synchronous Deferreds" ==> "Create a synchronous Deferred"
 * the ```test_reprWithChaining``` docstring has the docstring of the test above it but is about waiting and not results.
 * L{Deferred} markup disappears after ```test_nestedAsynchronousChainedDeferredsWithExtraCallbacks```, if you want that consistency.
 * in defer.py: the Deferred class docstring references doc/howto/defer.html, which I don't believe exists anymore.

Other than that, musing on defer.py for a while I couldn't come up with basic cases that were missing coverage - looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

(In [28478]) Respond to jesstess's comments (refs [#4300](https://github.com/twisted/twisted/issues/4300))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set status to `closed`|
|-|-|

(In [28480]) Merge deferred-unit-tests-4300

 * Author: jml, exarkun
 * Reviewer: jesstess
 * Fixes [#4300](https://github.com/twisted/twisted/issues/4300)

Add a bunch of tests and a couple of debugging affordances to Deferred.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `reopened`|
|-|-|

Reverted in r28814:

Revert r28480 (merge of deferred-unit-tests-4300)

The change to `Deferred.__str__` made the CPU and memory requirements
for this method (at least) quadratic on the length of the Deferred
chain.  This makes it impossible to compute the string representation
for many Deferreds, and unreasonably expensive for many others.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29554]) Adjust the new repr to only followed a Deferred chain one link

refs [#4300](https://github.com/twisted/twisted/issues/4300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I changed the new repr code to only go one step in the chain.  It's probably possible to actually render the full chain in only O(N) time, but I didn't bother.  We can try that out at some future point.  I mainly want to get this merged before [#4593](https://github.com/twisted/twisted/issues/4593) because of the extra test coverage.

I made a few other minor changes, but nothing too surprising.  I think the branch looks good overall now (as it did before, with just the minor issue of the repr performance).  I'd like another set of eyes though, since I just made some changes and I also paired with jml a bit when the branch was first being worked on.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

 * Please make a full sentence in the NEWS fragment.
 * flake: twisted/test/test_defer.py:11: 'util' imported but unused
 * It seems test_asynchronousImplicitErrorChain should return a Deferred, considering what's said in the docstring.
 * test_doubleAsynchronousImplicitChaining ought to have a better docstring.
 * the comments of _NO_RESULT and _chainedTo can probably be moved in the module and class docstring, respectively.

Nothing major, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [29805]) Expand the test_doubleAsynchronousImplicitChaining docstring


refs [#4300](https://github.com/twisted/twisted/issues/4300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [29806]) Change the docstring for test_asynchronousImplicitErrorChain to make it clear that gc interaction is not being tested directly here

refs [#4300](https://github.com/twisted/twisted/issues/4300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [29807]) Document _NO_RESULT like anything else

refs [#4300](https://github.com/twisted/twisted/issues/4300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> Please make a full sentence in the NEWS fragment. 

Since it was a misc file, it should have been empty.  So I emptied it.

> flake: twisted/test/test_defer.py:11: 'util' imported but unused 

Oops.  It looks like this was introduced when `Deferred.setTimeout` was removed.  I think I'll leave it alone for now because it's still used in the version of `defer.py` in the branch, so removing it would complicate testing.

> It seems test_asynchronousImplicitErrorChain should return a Deferred, considering what's said in the docstring. 

I adjusted the docstring.  The gc interaction is tested elsewhere, so this test was claiming to be a little broader than it needed to be.  It verifies that the result of the deferred is None, not a failure, which seems good enough to me.

> test_doubleAsynchronousImplicitChaining ought to have a better docstring. 

But it's so complete, concise, and expressive!  It's the best docstring ever!  Yarhh, fine, I expanded it a bit.

> the comments of _NO_RESULT and _chainedTo can probably be moved in the module and class docstring, respectively. 

Hm.  <IRC conversation>.  Okay.  The conclusion is that docstrings are better than comments because tools (like pydoctor) can deal with them.  And we document everything else in docstrings.  So, fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [29808]) Branching to 'deferred-unit-tests-4300-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [29811]) Document _chainedTo

refs [#4300](https://github.com/twisted/twisted/issues/4300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

Woops, I needed to merge forward anyway to get a clean test run, so I fixed that pyflakes warning after all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `closed`|
|-|-|

(In [29812]) Merge deferred-unit-tests-4300-2

Author: jml, exarkun
Reviewer: jesstess, therve
Fixes: [#4300](https://github.com/twisted/twisted/issues/4300)

Add a bunch of tests and a couple of debugging affordances to Deferred.
This re-merge fixes a quadratic complexity issue with the new `__repr__`
for Deferred.  Only a single step of chaining is now represented in the
repr string for a Deferred.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
